pkg/bdb/wrapper.go

```diff
@@ -22,18 +24,40 @@ func (w *Wrapper) Get(key []byte) ([]byte, error) {
 		}
 	})
 	if err != nil {
-		if err == badger.ErrKeyNotFound {
-			return nil, core.ErrKeyAbsent(key).Err()
-		}
 		return nil, err
 	}
-	return res, nil
+
+	document := &dstk.DcDocument{}
+	err = proto.Unmarshal(res, document)
+	return document, err
 }
 
 // thread safe
-func (w *Wrapper) Put(key []byte, value []byte, ttlSeconds float32) error {
+func (w *Wrapper) Put(key []byte, document *dstk.DcDocument, ttlSeconds float32) error {
 	return w.Update(func(txn *badger.Txn) error {
-		entry := badger.NewEntry(key, value).WithTTL(time.Duration(ttlSeconds) * time.Second)
+		// We need to fetch current document to compare etag.
+		currentDocument, err := w.Get(key)
+		newDocument := err == badger.ErrKeyNotFound
+		if err != nil && !newDocument {
+			return err
+		}
+		if !newDocument && (document.GetEtag() != currentDocument.GetEtag()) {
+			return badger.ErrConflict
+		}
+
+		// Set a new etag.
+		randomEtag, err := uuid.NewRandom()
```

Comment:
During last discussion i did insist on using random, but random string wont work. We need some notion of ordering for verification of transactions. Can you use some monotonic sequence instead. Again with ability to mock for verification

Reply:
Done.